hedera-node/hedera-token-service-impl/src/main/java/com/hedera/node/app/service/token/impl/handlers/transfer/customfees/CustomFeeAssessor.java

```diff
@@ -0,0 +1,97 @@
+import com.hedera.hapi.node.token.CryptoTransferTransactionBody;
+import com.hedera.node.app.spi.workflows.HandleContext;
+import edu.umd.cs.findbugs.annotations.NonNull;
+import javax.inject.Inject;
+
+/**
+ * Assesses custom fees for a given crypto transfer transaction.
+ */
+public class CustomFeeAssessor {
+    private final CustomFixedFeeAssessor fixedFeeAssessor;
+    private final CustomFractionalFeeAssessor fractionalFeeAssessor;
+    private final CustomRoyaltyFeeAssessor royaltyFeeAssessor;
+    private int initialNftChanges = 0;
+
+    @Inject
+    public CustomFeeAssessor(
+            @NonNull final CustomFixedFeeAssessor fixedFeeAssessor,
+            @NonNull final CustomFractionalFeeAssessor fractionalFeeAssessor,
+            @NonNull final CustomRoyaltyFeeAssessor royaltyFeeAssessor,
+            @NonNull final CryptoTransferTransactionBody op) {
+        this.fixedFeeAssessor = fixedFeeAssessor;
+        this.fractionalFeeAssessor = fractionalFeeAssessor;
+        this.royaltyFeeAssessor = royaltyFeeAssessor;
+        initialNftChanges = numNftTransfers(op);
+    }
+
+    private int numNftTransfers(final CryptoTransferTransactionBody op) {
+        final var tokenTransfers = op.tokenTransfersOrElse(emptyList());
+        var nftTransfers = 0;
+        for (final var xfer : tokenTransfers) {
+            nftTransfers += xfer.nftTransfersOrElse(emptyList()).size();
+        }
+        return nftTransfers;
+    }
+
+    public void assess(
+            final AccountID sender,
+            final CustomFeeMeta feeMeta,
+            final int maxTransfersSize,
+            final AccountID receiver,
+            final AssessmentResult result,
+            final HandleContext ctx) {
+        fixedFeeAssessor.assessFixedFees(feeMeta, sender, result);
+        validateBalanceChanges(result, maxTransfersSize);
+
+        // A FUNGIBLE_COMMON token can have fractional fees but not royalty fees.
+        // A NON_FUNGIBLE_UNIQUE token can have royalty fees but not fractional fees.
+        // So check token type and do further assessment
+        if (feeMeta.tokenType().equals(TokenType.FUNGIBLE_COMMON)) {
+            fractionalFeeAssessor.assessFractionalFees(feeMeta, sender, result);
+        } else {
+            royaltyFeeAssessor.assessRoyaltyFees(feeMeta, sender, receiver, result, ctx);
+        }
+        validateBalanceChanges(result, maxTransfersSize);
+    }
+
+    private void validateBalanceChanges(final AssessmentResult result, final int maxTransfersSize) {
+        var inputFungibleTransfers = 0;
+        var newFungibleTransfers = 0;
+        for (final var entry : result.getInputTokenAdjustments().entrySet()) {
+            inputFungibleTransfers += entry.getValue().size();
+        }
+        for (final var entry : result.getHtsAdjustments().entrySet()) {
+            newFungibleTransfers += entry.getValue().size();
+        }
+        final var balanceChanges = result.getHbarAdjustments().size()
```

Comment:
It's quite nice to be able to see the count of balance changes derived so explicitly. 🤩